File: frontend/src/api/themes.js
import client from './client.js'

export const listThemes = () => client.get('/themes').then(r => r.data)
export const getPresets = () => client.get('/themes/presets').then(r => r.data)
export const getTheme = (id) => client.get(`/themes/${id}`).then(r => r.data)
export const createTheme = (data) => client.post('/themes', data).then(r => r.data)
export const updateTheme = (id, data) => client.put(`/themes/${id}`, data).then(r => r.data)
export const deleteTheme = (id) => client.delete(`/themes/${id}`).then(r => r.data)
export const extractFromLogo = (file) => {
  const form = new FormData()
  form.append('file', file)
  return client.post('/themes/extract-from-logo', form, { headers: { 'Content-Type': 'multipart/form-data' } }).then(r => r.data)
}

File: backend/app/routers/themes.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from sqlalchemy.orm import Session
from typing import List
from app.db import get_db
from app import models
from app.auth import get_current_user
from app.schemas import ThemeCreate, ThemeUpdate, ThemeOut
from app.theme_registry import get_all_presets

router = APIRouter()


@router.get("/presets")
def list_presets():
    return get_all_presets()


@router.get("", response_model=List[ThemeOut])
def list_themes(db: Session = Depends(get_db), current_user=Depends(get_current_user)):
    return db.query(models.Theme).all()


@router.post("", response_model=ThemeOut)
def create_theme(data: ThemeCreate, db: Session = Depends(get_db), current_user=Depends(get_current_user)):
    obj = models.Theme(**data.model_dump())
    db.add(obj)
    db.commit()
    db.refresh(obj)
    return obj


@router.get("/{id}", response_model=ThemeOut)
def get_theme(id: int, db: Session = Depends(get_db), current_user=Depends(get_current_user)):
    obj = db.query(models.Theme).filter_by(id=id).first()
    if not obj:
        raise HTTPException(404, "Theme not found")
    return obj


@router.put("/{id}", response_model=ThemeOut)
def update_theme(id: int, data: ThemeUpdate, db: Session = Depends(get_db), current_user=Depends(get_current_user)):
    obj = db.query(models.Theme).filter_by(id=id).first()
    if not obj:
        raise HTTPException(404, "Theme not found")
    for k, v in data.model_dump(exclude_none=True).items():
        setattr(obj, k, v)
    db.commit()
    db.refresh(obj)
    return obj


@router.delete("/{id}")
def delete_theme(id: int, db: Session = Depends(get_db), current_user=Depends(get_current_user)):
    obj = db.query(models.Theme).filter_by(id=id).first()
    if not obj:
        raise HTTPException(404, "Theme not found")
    if obj.is_preset:
        raise HTTPException(400, "Cannot delete preset theme")
    db.delete(obj)
    db.commit()
    return {"deleted": True}


@router.post("/extract-from-logo")
async def extract_from_logo(file: UploadFile = File(...), current_user=Depends(get_current_user)):
    """Extract dominant colors from a logo image."""
    from app.services.theme_service import extract_colors_from_image
    content = await file.read()
    colors = extract_colors_from_image(content)
    return {"extracted_colors": colors}
